### PR TITLE
fix(jdbc-sink): Handle JDBC connection error by recreating the connection

### PR DESCRIPTION
**Pull request by StrikeW**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

- Handle JDBC connection error by recreate the connection and write the batch again, if fail to recreate we will throw an exception
- Refactor some code.

related: https://github.com/risingwavelabs/risingwave/issues/13170
close https://github.com/risingwavelabs/risingwave/issues/13244

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by StrikeW** [comment]
No performance degradation compared to main. Sink 1.5M rows tpch orders table to postgres.

image

**Comment by StrikeW** [comment]
PTAL of this PR which should go into v1.4 to improve stability. cc @hzxa21 @liurenjie1024

**Comment by hzxa21** [comment]
@wenym1 and I had an offline discussion about whether using connection pool is appropriate here. Overhead aside, we have two concerns:
1. Connection can be dropped or closed at any given point. Try to get a connection on `write` does solve the issue when the connection is dropped in between `write` call but If the connection is closed by external system during `write`, the current approach won't work.
2. Connection pool is meant for the cases when you have many concurrent short-lived requests wanting to reuse connections. However, in the current JDBC sink implementation, the connection is long-lived and the queries are sequential. That made me think whether connection pool is the right solution here.

After #13258, the iterator becomes a iterable that can be reset. How about try recreating the connection (w/o connection pool) and retrying `write` by reseting the iterable on exception? @wenym1 What do you think?

**Comment by wenym1** [comment]
> After https://github.com/risingwavelabs/risingwave/pull/13258, the iterator becomes a iterable that can be reset. How about try recreating the connection (w/o connection pool) and retrying write by reseting the iterable on exception? @wenym1 What do you think?

Agree. I think it's better to recreate a new connection when we get some error on write. Connection pool is dedicated for sharing connection for db access with high concurrency and short use time, which is not the case for the jdbc sink.

**Comment by liurenjie1024** [comment]
I prefer to use connection pool to help managing connection for us, since managing jdbc connection doesn't sound a trivial task for me. But why we need a pool per executor? I think sharing one connection pool per instance would be enough.

**Comment by hzxa21** [comment]
> I prefer to use connection pool to help managing connection for us, since managing jdbc connection doesn't sound a trivial task for me. But why we need a pool per executor? I think sharing one connection pool per instance would be enough.

+1. Maintaining one connection pool per instance sounds a good optimization to me. I view it orthogonal to error handling and prepare statement retry though.

**Comment by StrikeW** [comment]
> Connection can be dropped or closed at any given point. Try to get a connection on write does solve the issue when the connection is dropped in between write call but if the connection is closed by external system during write, the current approach won't work.

I think this conclusion is not true. As I have done some experiments, first I terminate all the established connections  by `pg_terminate_backend`, then I insert new rows to the RW table, the rows can successfully go into the PG table, and the connection can be re-created by the connection pool. cc @hzxa21 

image


The connection pool is capped to 1 connection, and we can use it for connection management in our use case.

**Comment by StrikeW** [comment]
> if the connection is closed by external system **during** write, the current approach won't work.

Oh, I think your point is **during** write. The connection pool may not handle this case, so basically you mean the JDBC sink needs to **ensure** the write eventually success, am I got you right? cc @hzxa21 @wenym1

**Comment by liurenjie1024** [comment]
> The connection pool is capped to 1 connection, and we can use it for connection management in our use case.

Connection pool may contains complex states such as time, task pool.

**Comment by StrikeW** [comment]
> > The connection pool is capped to 1 connection, and we can use it for connection management in our use case.
> 
> Connection pool may contains complex states such as time, task pool.

So you mean it is costly to maintain it, right? I agree to you it would be cost to create a large number of connection pools. I will refactor the code to use connection instead.

**Comment by StrikeW** [comment]
Refactored. PTAL again cc @hzxa21 @wenym1

**Review by wenym1 — COMMENTED** [review]
Seems that we are modifying the interface of `SinkWriterV1`. It's better to reimplement `SinkWriter` for the `JdbcSink` instead of changing the original one. The `SinkWriterV1` is for compatibility for sinks already implemented.

File: java/connector-node/risingwave-sink-jdbc/src/main/java/com/risingwave/connector/JDBCSink.java [hunk 1]
```diff
@@ -167,133 +124,222 @@ private static List<String> getPkColumnNames(
...
+                        }
+                    } else {
+                        jdbcStatements.prepareInsert(row);
+                    }
+                }
+                // Execute staging statements after all rows are prepared.
+                jdbcStatements.execute();
+                break;
+            } catch (SQLException e) {
+                LOG.error("Failed to execute JDBC statements, retried {} times", retryCount, e);
```
Comment (wenym1) [inline java/connector-node/risingwave-sink-jdbc/src/main/java/com/risingwave/connector/JDBCSink.java:152]: Seems that we are endlessly retrying? Shall we set a max retry count? And is there a way to check whether the SQLException is caused by network connection or some other data failure (such key len exceed limit)? It's better to create the connection when the exception is caused by network error.

**Comment by StrikeW** [comment]
> Seems that we are modifying the interface of `SinkWriterV1`. It's better to reimplement `SinkWriter` for the `JdbcSink` instead of changing the original one. The `SinkWriterV1` is for compatibility for sinks already implemented.

I think change from `Iterator` to `Iterable` doesn't break existing sinks and reimplement `SinkWriter` for the `JdbcSink` is out of the scope of this PR, we should left this to a separate PR.

**Comment by wenym1** [comment]
> Iterator to Iterable doesn't break existing sinks

Many unnecessary changes in other sinks are included in this PR. 

The `SinkWriterV1` is deprecated and should not be changed anymore.It's simple to reimplement the jdbc sink with the new SinkWriter, since most logic is in the `write` method, and the only difference between `SinkWriterV1` and `SinkWriter` is changing `Iterator` to `Iterable`, which is exactly the feature we expect. Therefore it's better to just reimplement jdbc sink with the new interface, and the PR can be much cleaner.

**Comment by wenym1** [comment]
In https://github.com/risingwavelabs/risingwave/pull/13190, I just reimplemented the jdbc sink with the new interface in commit https://github.com/risingwavelabs/risingwave/pull/13190/commits/85340c8593b8e344749cc78652daf2d0b8d977ff. The line of change is less than 30. Feel free to cherrypick the commit to be included in this PR first.

**Review by wenym1 — APPROVED** [review]

File: java/connector-node/risingwave-sink-jdbc/src/main/java/com/risingwave/connector/JDBCSink.java [hunk 2]
```diff
@@ -167,168 +127,250 @@ private static List<String> getPkColumnNames(
...
+    @Override
+    public void write(Iterable<SinkRow> rows) {
+        final int maxRetryCount = 4;
+        int retryCount = 0;
+        while (true) {
+            try {
+                // fill prepare statements with parameters
+                for (SinkRow row : rows) {
+                    if (row.getOp() == Data.Op.UPDATE_DELETE) {
+                        updateFlag = true;
```
Comment (hzxa21) [inline java/connector-node/risingwave-sink-jdbc/src/main/java/com/risingwave/connector/JDBCSink.java:139]: Do we need to record the value of `updateFlag` before entring the for loop and reset it when retry happens?

File: java/connector-node/risingwave-sink-jdbc/src/main/java/com/risingwave/connector/JDBCSink.java [hunk 3]
```diff
@@ -167,168 +127,250 @@ private static List<String> getPkColumnNames(
...
-                case INSERT:
-                    jdbcDialect.bindUpsertStatement(preparedStmt, conn, getTableSchema(), row);
-                    break;
-                case UPDATE_INSERT:
-                    if (!updateFlag) {
-                        throw Status.FAILED_PRECONDITION
-                                .withDescription("an UPDATE_DELETE should precede an UPDATE_INSERT")
-                                .asRuntimeException();
+                try {
+                    if (!conn.isValid(30)) { // 30 seconds timeout
```
Comment (hzxa21) [inline java/connector-node/risingwave-sink-jdbc/src/main/java/com/risingwave/connector/JDBCSink.java:166]: How is 30seconds timeout chosen? Will it be too long? I saw that the default isValid check timeout for JDBC conn is 3s.

Comment (hzxa21) [inline java/connector-node/risingwave-sink-jdbc/src/main/java/com/risingwave/connector/JDBCSink.java:166]: We will continue to retry if `conn.isValid` returns true. Is this expected?

File: java/connector-node/risingwave-sink-jdbc/src/main/java/com/risingwave/connector/JDBCSink.java [hunk 4]
```diff
@@ -167,168 +127,250 @@ private static List<String> getPkColumnNames(
...
+                            jdbcStatements.prepareUpsert(row);
+                        }
+                    } else {
+                        jdbcStatements.prepareInsert(row);
+                    }
+                }
+                // Execute staging statements after all rows are prepared.
+                jdbcStatements.execute();
+                break;
+            } catch (SQLException e) {
```
Comment (hzxa21) [inline java/connector-node/risingwave-sink-jdbc/src/main/java/com/risingwave/connector/JDBCSink.java:155]: Have we verified that a SQLExection, not other exceptions like RuntimeException, will be thrown on connection close?

**Review by hzxa21 — APPROVED** [review]

**Pull request merged**